module/zfs/dbuf.c

```diff
@@ -1234,11 +1224,44 @@ dbuf_read_impl(dmu_buf_impl_t *db, zio_t *zio, uint32_t flags)
 		spa_log_error(db->db_objset->os_spa, &zb);
 		zfs_panic_recover("unencrypted block in encrypted "
 		    "object set %llu", dmu_objset_id(db->db_objset));
+		DB_DNODE_EXIT(db);
+		mutex_exit(&db->db_mtx);
 		return (SET_ERROR(EIO));
 	}
 
+	/*
+	 * If we are doing a decrypting read of a block, make sure we
+	 * have decrypted the dnode associated with it so that the current
+	 * bp's MAC can be verified.
+	 */
+	if (abuf != NULL && dn->dn_objset->os_encrypted &&
+	    DMU_OT_IS_ENCRYPTED(dn->dn_bonustype) &&
```

Comment:
are the `os_encrypted` and `OT_IS_ENCRYPTED` checks necessary, given that we are checking `arc_is_encrypted`?

Reply:
Technically no. These just enable us to quickly bail without doing a function call for unencrypted datasets and the MOS. Would you like me to remove it?

Reply:
It's not like the arc_is_encrypted() call is slow, is it?  I'd say just do the checks that make the code easiest to understand.